clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7623,14 +7623,17 @@ SYCLIntelFPGAMaxConcurrencyAttr *Sema::MergeSYCLIntelFPGAMaxConcurrencyAttr(
   // Check to see if there's a duplicate attribute with different values
   // already applied to the declaration.
   if (const auto *DeclAttr = D->getAttr<SYCLIntelFPGAMaxConcurrencyAttr>()) {
-    const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr());
-    const auto *MergeExpr = dyn_cast<ConstantExpr>(A.getNThreadsExpr());
-    if (DeclExpr && MergeExpr &&
-        DeclExpr->getResultAsAPSInt() != MergeExpr->getResultAsAPSInt()) {
-      Diag(DeclAttr->getLoc(), diag::warn_duplicate_attribute) << &A;
-      Diag(A.getLoc(), diag::note_previous_attribute);
+    if (const auto *DeclExpr =
+            dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr())) {
+      if (const auto *MergeExpr = dyn_cast<ConstantExpr>(A.getNThreadsExpr())) {
+        if (DeclExpr->getResultAsAPSInt() != MergeExpr->getResultAsAPSInt()) {
+          Diag(DeclAttr->getLoc(), diag::warn_duplicate_attribute) << &A;
+          Diag(A.getLoc(), diag::note_previous_attribute);
+        }
+        // Do not add a duplicate attribute.
+        return nullptr;
```

Comment:
I'm not sure why this change was made. I feel the old code was  easier to read. You've also slightly changed behavior here right? The return was hit before if there was an existing attribute. Now the return is inside - 

```
 if (const auto *DeclExpr =
            dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr()))
```

Reply:
Sorry for late response. Thanks @elizabethandrews for reviews.

I found a bug with the original implementation while working on splitting tests (AST vs Sema tests).

Template case(below) for duplicate attributes does not work well with the original implementation: Test compiles fine without any diagnostic. I have made similar changes like other function attributes to handle template cases for duplicate attribute values with max_concurrency attribute properly. 

```
template <int size>
[[intel::max_concurrency(10)]] void func11(); // expected-note {{previous attribute is here}}
template <int size>
[[intel::max_concurrency(size)]] void func11() {} // expected-warning {{attribute 'max_concurrency' is already applied with different arguments}}

void checkTemplates() {
  
  func11<20>(); // expected-note {{in instantiation of function template specialization 'func11<20>' requested here}}
}
```